### PR TITLE
Form block: Add color & spacing design tool

### LINKED ISSUE #19951: Form Block: custom CSS Classes do not load on the Front-end
### Steps to reproduce the behaviour

1. Add a Contact Form Block (or any other Form Block - RSVP etc)
2. Go to Block > Advanced > Additional CSS Classes and add any Custom Class here.
3. Publish the page. 
4. Open the Page on the Front-end and check the Source-code for the Additional Class.

#### What I expected to happen

The Additional CSS Class should load on the Front-end too

#### What actually happened

The Additional CSS Class does not load

#### Browser / OS version

Browser: Chrome Version 90.0.4430.212
OS: MacOS Big Sur 11.4

#### Screenshot / Video: If applicable, add screenshots to help explain your problem.

https://user-images.githubusercontent.com/16253818/119524362-dc543580-bd7d-11eb-8dc9-3d5dae24a245.mp4

### PR DESCRIPTION
**Pull request by ivan-ottinger**
Fixes #19951 

The proposed changes add Color and Spacing (Padding) features to the Form block:

 Before | After |
| ------------- | ------------- |
| Markup on 2022-01-26 at 15:11:28 | Markup on 2022-02-03 at 14:21:50 |

The changes also address the following issue: https://github.com/Automattic/jetpack/issues/19951 (missing custom CSS class).

**Screen recording:**

https://user-images.githubusercontent.com/25105483/152351833-76c5c563-acc9-4c1e-8604-0436090fc29f.mp4

#### Changes proposed in this Pull Request:

- add wrapper `div` that will hold all the necessary classes and styles
- add available block `supports`: color and spacing
- deprecate the current block version as version v2 and adjust fixtures

#### Jetpack product discussion
This PR is a part of the _Expanding Jetpack Block design tooling for Full Site Editing_ project: pdDOJh-6-p2.

#### Does this pull request change what data or activity we track or use?
No, it does not.

#### Testing instructions:
Please note that the Form block is used by many users and sites. It is very important we test this PR thoroughly to make sure everything is working as expected.

##### Main test
1. Create a new post or a page with the Form block (if the block isn't displaying in the block inserter, you may need to navigate to `http://localhost/wp-admin/admin.php?page=jetpack_modules` and activate the `Contact Form` module)
2. There should be a new set of features in the block sidebar: Color and Dimensions / Padding
3. Experiment with the new features and observe whether the change applies correctly in the back-end
4. Save the changes and review if the settings are correctly rendering in the front-end as well
5. We can test the change with multiple themes (block-based, classic). Not all themes support all features. If a theme doesn't support a specific feature, this feature won't be displayed in the block sidebar.

For example, themes like Storefront and Dara don't support block Dimensions / Padding and therefore it won't display in the block sidebar. Theme like Hever, Maywood or Geologist support both: Color and Dimensions / Padding.

6. It shouldn't be possible to reproduce the https://github.com/Automattic/jetpack/issues/19951 issue.

##### Backward compatibility test
1. **Before applying the changes in this PR**, create a new post or a page with the Form block. You can create multiple kinds of Form block and customize them (change field names, adjust block settings in the sidebar, ...):

Markup on 2022-01-26 at 15:18:28

2. Adjust the block settings and save the page
3. Apply the code changes from this PR / patch
4. Reload the block editor and make sure the block is still working as expected. The block should validate correctly
5. Make additional changes to the block and save them
6. Review if the block is rendering correctly in both back-end and front-end

Apart from testing how the block looks like, we should also test block functionality: e.g. if the form can be properly submitted.

##### WPCOM test
We need to perform all tests above on the WPCOM sandbox as well.

---

Useful links:

* [Form block documentation (Jetpack)](https://jetpack.com/support/contact-form/)
* [Form block documentation (WPCOM)](https://wordpress.com/support/wordpress-editor/blocks/form-block/)

### CODE REVIEW

**Comment by yansern** [comment]
On our discussion earlier on failing deprecated fixture tests, `originalContent` containing empty string is correct and there isn't a need to meddle with the value of the object. The `originalContent` basically holds the HTML value of what the fixture file looks like before it is parsed. In our case, the deprecated fixture file only contains HTML comments, which results in empty `originalContent`.

I looked into the following files to try to roughly understand what's going on under the hood.
- extensions/shared/test/block-fixtures.js
- packages/block-serialization-default-parser/parser.php (gutenberg)
- packages/blocks/src/api/parser/index.js (gutenberg)

**Comment by yansern** [comment]
> The change also addresses the following issue: #19951 (missing custom CSS class).

This is as a result of applying block props (class/style) attributes into the wrapper div, which is what is needed to get design tools to work.

**Review by yansern — DISMISSED** [review]
In summary:
- A wrapper div is added where class and style attributes are applied using `useBlockProps`.
- The `deprecated.js` file contains V1 and V2. V1 is for the earlier attributes format, V2 is for the old `save()` method where a div wrapper is not added.
- New fixture HTML files for V2 and V3 in `blocks/contact-form/test/fixtures/*` are added, the results are regenerated to match the output of V3.

**Comment by ivan-ottinger** [comment]
> The `originalContent` basically holds the HTML value of what the fixture file looks like before it is parsed. In our case, the deprecated fixture file only contains HTML comments, which results in empty `originalContent`.

Thanks for digging into this Yan.

When I was reading the [doc on regenerating fixtures](https://github.com/Automattic/jetpack/blob/master/projects/plugins/jetpack/extensions/shared/test/block-fixtures.md):

> The .json contains the updated attributes and inner blocks after the deprecation process. These are then passed through the current block save method to see if a valid block is produced.

> In the case of deprecations, the content of the .json and .serialized.html files should match the content produced by the latest version of the block save method, ie. they should mimic what would happen to a block's content when a deprecated version is loaded in the editor and migrated to be compatible with the new save method.

It made me think that the `.json` file of the deprecated version should be identical to the `.json` file of the current block version.

In other words, I expected that even when the deprecated seed `.html` file doesn't contain the wrapper div, the fixture will assume the `.json` output should hold that extra `div` in `originalContent` already:

```
"originalContent": "<div class=\"wp-block-jetpack-contact-form noodles\">\n\n\n    \n\n    \n</div>"`.
```

But as you mentioned in your comment, this is no the case.

From what I understand now, the deprecated `.json` file contains the values of the block **_before_** (not **__after_**) the deprecated block content turns into the updated block content:

https://user-images.githubusercontent.com/25105483/151766864-9321b8b0-3d58-40a3-b63b-daf5d2d44737.mp4

👍🏼

**Comment by ivan-ottinger** [comment]
I have also tested the changes with the Jetpack CRM and Creative Mail integrations.

→ Worked as expected, the Form blocks that had been created before the changes from this PR were applied migrated to the new Form block version correctly. Integrations worked after the block migration (the data from the form were properly recorded to Jetpack CRM and Creative Mail respectively).

**Review by jeherve — COMMENTED** [review]
I can see the new options in the UI, but the typography options do not seem to do anything on my end (I'm using Chrome and Twenty Twenty:

https://user-images.githubusercontent.com/426388/152194078-5eadc5fa-4033-486c-af15-0f10cedd0940.mov

**Comment by ivan-ottinger** [comment]
> I can see the new options in the UI, but the typography options do not seem to do anything on my end (I'm using Chrome and Twenty Twenty:
> 
>  Screen.Recording.2022-02-02.at.17.20.32.mov

Thank you for your review, Jeremy! I didn't notice that issue before.

---

It turns out the Twenty Twenty theme stylesheet is overriding the styles set by the user in the editor to `font-size: 1.6rem;` (in `style.css`, front end):

```css
label {
	display: block;
	font-size: 1.6rem;
	font-weight: 400;
	margin: 0 0 0.5rem 0;
}
```

and also in the editor (`editor-style-block.css`):

```css
.editor-styles-wrapper label {
	font-size: 15px;
	font-weight: 600;
}
```

It looks similar to the font-size issue related to the Table block that has been reported with the Twenty Twenty theme last December: https://github.com/Automattic/themes/issues/5205

I have now also noticed similar behavior with the Twenty Twenty-One theme.

---

The first idea I had to address this issue when writing this comment was to keep the Typography features available on the Form block and report bug for each theme. But thinking about this further, it might be better to drop the support for Typography (on the Form block). At least for now.

The reason for this is that at the [current state](https://github.com/WordPress/gutenberg/blob/add/styles-intro/docs/explanations/architecture/styles.md#current-limitations-of-the-block-supports-api) of "Supports API", each design tool applies to the whole block (in form of `style` and `class` attributes of the block wrapper tag).

This means that the same font-size would apply to all elements of the Form block, including labels, inputs, text areas, etc.... As a result, I suspect that more inconsistencies could arise with other themes.

Also, from the user's perspective, I think the better experience would be if the user could set the font-size for each form element separately (e.g. one font-size for labels, other font-size for input fields). This would require more advanced changes to the Form block.

If a real need to have the Typography tools available on the Form block arises, we can still add it in the future.

---

To summarize, I think it is the best to remove Typography tools from the proposed Form block changes for now.

What do you think? Also pinging @yansern if you would like to share your thoughts. Cheers.

**Comment by jeherve** [comment]
> I think it is the best to remove Typography tools from the proposed Form block changes for now.

Yes, I would agree. There are so many different types of inputs in the form block, I do not think it is realistic to expect a single typography option that would manage all those things properly.

**Comment by yansern** [comment]
> What do you think? Also pinging @yansern if you would like to share your thoughts. Cheers.

Yes, I think it make sense to remove Typography as it needs to be more granular.

**Review by yansern — APPROVED** [review]
Rebased & retested everything. This is good to go.

**Review by jeherve — COMMENTED** [review]
The new alignment options do not seem to work for me with Quadrat. The class is properly added to the block container, but since it's contained within another div that only has an id with the contact form ID, no alignment changes happen.

image

**Comment by ivan-ottinger** [comment]
> The new alignment options do not seem to work for me with Quadrat. The class is properly added to the block container, but since it's contained within another div that only has an id with the contact form ID, no alignment changes happen.

Thank you Jeremy.

It doesn't seem to work with any other theme either - as they don't have effect on the block structure and how the `divs` are nested. 

I have looked into this a bit deeper and it doesn't seem that straightforward to fix. Also, it would require additional changes to the CSS, especially for the Full alignment. I vote to work on the Form block alignment tool in a separate PR so we can get the color and spacing tools live finally.

Even if we fixed the alignment tool in this PR, I would say it would be too many changes for a single PR.

What do you think? Also pinging Yan @yansern - if you would like to share your point of view.

**Comment by yansern** [comment]
> What do you think? Also pinging Yan @yansern - if you would like to share your point of view.

Okay, let's remove the alignment tool so we can get a move on. (I realized it was me who added it. Now I don't recall if it was working for me.)

**Comment by ivan-ottinger** [comment]
> Okay, alignment removed. Sorry for the unnecessary hold up! :)

No worries at all! :)

**Comment by ivan-ottinger** [comment]
Rebased and tested again. Works well. 👍🏼

**Review by samiff — APPROVED** [review]
Tested well for me. I added the old unserialized HTML versions to the editor and was able to see those validate/update to the new block version. Tried setting block options across a few different themes and looked good.
